python/cuml/cuml/internals/interop.py

```diff
@@ -0,0 +1,311 @@
+    CPU-based counterpart.
+
+    Subclasses should define:
+
+    - `_cpu_class_path`
+    - `_params_from_cpu`
+    - `_params_to_cpu`
+    - `_attrs_from_cpu`
+    - `_attrs_to_cpu`
+
+    In return they get ``as_sklearn``/``from_sklearn`` methods, and can also
+    be used as the basis for a proxy estimator in ``cuml.accel``.
+    """
+
+    # The import path to use to import the CPU model class
+    _cpu_class_path: str
+
+    @classproperty
+    def _cpu_class(cls):
+        """The CPU class that corresponds to this GPU model"""
+        module, _, name = cls._cpu_class_path.rpartition(".")
+        return getattr(import_module(module), name)
+
+    @classmethod
+    def _params_from_cpu(cls, model) -> dict[str, Any]:
+        """Get parameters to use to instantiate a GPU model from a CPU model.
+
+        Parameters
+        ----------
+        model
+            The CPU model, of the same type as ``self._cpu_class``.
+
+        Returns
+        -------
+        dict
+            A mapping of keyword arguments that may be used to instantiate ``cls``
+            to create an equivalent GPU model.
+
+        Raises
+        ------
+        UnsupportedOnGPU
+            If one or more hyperparameters are unsupported by the GPU model.
+        """
+        raise NotImplementedError
+
+    def _params_to_cpu(self) -> dict[str, Any]:
+        """Get parameters to use to instantiate a CPU model from a GPU model.
+
+        Returns
+        -------
+        dict
+            A mapping of keyword arguments that may be used to instantiate
+            ``self._cpu_class`` to create an equivalent CPU model.
+
+        Raises
+        ------
+        UnsupportedOnCPU
+            If one or more hyperparameters are unsupported by the CPU model.
+        """
+        raise NotImplementedError
+
+    def _attrs_from_cpu(self, model) -> dict[str, Any]:
+        """Get attributes to set on ``self`` from a fit CPU model.
+
+        The callers of this method check that ``model`` is of the correct
+        type and has been fit, no need to check that in an implementation.
+
+        The base class handles common metadata attributes, implementations should
+        be sure to call ``super()._attrs_from_cpu()`` to include those in the output.
+
+        Parameters
+        ----------
+        model
+            The CPU model, of the same type as ``self._cpu_class``.
+
+        Returns
+        -------
+        dict
+            A mapping of attributes to set on ``self`` to create a fit GPU model
+            from a fit CPU model.
+
+        Raises
+        ------
+        UnsupportedOnGPU
+            If one or more attributes are unsupported by the GPU model.
+        """
+        out = {}
+        for name in ["n_features_in_", "feature_names_in_"]:
+            try:
+                out[name] = getattr(model, name)
+            except AttributeError:
+                pass
+        return out
+
+    def _attrs_to_cpu(self, model) -> dict[str, Any]:
```

Comment:
This method shouldn't set things on `model` itself, the caller will do that, so why do we pass in `model`?

Reply:
Sometimes to properly recreate the CPU estimator we need to access its hyperparams to decide how to coerce the cuml state to the CPU state.

One example is the handling of `solver_` here: https://github.com/rapidsai/cuml/blob/926d044d72ac3ab95980d4bf449c017f6fd1c22d/python/cuml/cuml/linear_model/ridge.pyx#L257-L260. Because our mappings aren't 1:1, we need the original hyperparameter to pick a sane `solver_` value for the CPU output.